src/allmydata/scripts/tahoe_backup.py

```diff
@@ -313,14 +263,353 @@ def upload(self, childpath):
 
 def backup(options):
     bu = BackerUpper(options)
     return bu.run()
+
+
+def collect_backup_targets(root, listdir, filter_children):
+    """
+    Yield BackupTargets in a suitable order for processing (deepest targets
+    before their parents).
+    """
+    try:
+        children = listdir(root)
+    except EnvironmentError:
+        yield PermissionDeniedTarget(root, isdir=True)
+    except FilenameEncodingError:
+        yield FilenameUndecodableTarget(root, isdir=True)
+    else:
+        for child in filter_children(children):
+            assert isinstance(child, unicode), child
+            childpath = os.path.join(root, child)
+            if os.path.islink(childpath):
+                yield LinkTarget(childpath, isdir=False)
+            elif os.path.isdir(childpath):
+                child_targets = collect_backup_targets(
+                    childpath,
+                    listdir,
+                    filter_children,
+                )
+                for child_target in child_targets:
+                    yield child_target
+            elif os.path.isfile(childpath):
+                yield FileTarget(childpath)
+            else:
+                yield SpecialTarget(childpath)
+        yield DirectoryTarget(root)
+
+
+
+def run_backup(
+        warn,
+        upload_file,
+        upload_directory,
+        targets,
+        start_timestamp,
+        stdout,
+):
+    progress = BackupProgress(warn, start_timestamp, len(targets))
+    for target in targets:
+        # Pass in the progress and get back a progress.  It would be great if
+        # progress objects were immutable.  Then the target's backup would
+        # make a new progress with the desired changes and return it to us.
+        # Currently, BackupProgress is mutable, though, and everything just
+        # mutates it.
+        progress = target.backup(progress, upload_file, upload_directory)
+        print >>stdout, progress.report(datetime.datetime.now())
+    return progress.backup_finished()
+
+
+
+class FileTarget(object):
+    def __init__(self, path):
+        self._path = path
+
+
+    def __repr__(self):
+        return "<File {}>".format(self._path)
+
+    def backup(self, progress, upload_file, upload_directory):
+        try:
+            created, childcap, metadata = upload_file(self._path)
+        except EnvironmentError:
+            target = PermissionDeniedTarget(self._path, isdir=False)
+            return target.backup(progress, upload_file, upload_directory)
+        else:
+            assert isinstance(childcap, str)
+            if created:
+                return progress.created_file(self._path, childcap, metadata)
+            return progress.reused_file(self._path, childcap, metadata)
+
+
+
+class DirectoryTarget(object):
+    def __init__(self, path):
+        self._path = path
+
+
+    def __repr__(self):
+        return "<Directory {}>".format(self._path)
+
+
+    def backup(self, progress, upload_file, upload_directory):
```

Comment:
I *think* we mostly follow pep8 style, not Twisted style for newlines between things? (But, it's not consistent I see from a quick look, so ... )

Reply:
PEP8: "Python accepts the control-L (i.e. ^L) form feed character as whitespace; Many tools treat these characters as page separators, so you may use them to separate pages of related sections of your file."

:grin: 

I'll try to stick to 2/1 instead of 3/2.